indexer/osm_editor.cpp

```diff
@@ -0,0 +1,109 @@
+#include "osm_editor.hpp"
+
+#include "indexer/feature_decl.hpp"
+#include "indexer/index.hpp"
+
+#include "std/map.hpp"
+#include "std/set.hpp"
+
+namespace osm
+{
+// TODO(AlexZ, SergeyM): Use wrapper over XML (editor's "database" container).
+struct Editor::Impl
+{
+  // TODO(AlexZ): Use better structure(s) to store data, think about xml tree.
+  set<FeatureID> m_deletedFeatures;
+  map<FeatureID, FeatureType> m_editedFeatures;
+
+  Impl() { Load(); }
+  void Load()
+  {
+    // TODO(AlexZ): Initialize from xml file.
+  }
+  void Save()
+  {
+    // TODO(AlexZ): Save to xml file.
+  }
+};
+
+Editor::Editor() : m_impl(new Editor::Impl()) {}
+Editor::~Editor() { delete m_impl; }
+Editor::Impl & Editor::GetImpl()
+{
+  static Editor instance;
+  return *instance.m_impl;
+}
+
+bool Editor::IsFeatureDeleted(FeatureID const & fid)
+{
+  Impl & impl = GetImpl();
+  return impl.m_deletedFeatures.find(fid) != impl.m_deletedFeatures.end();
+}
+
+void Editor::DeleteFeature(FeatureID const & fid) { GetImpl().m_deletedFeatures.insert(fid); }
+static void ChangeName(StringUtf8Multilang & featureName, string const & newName)
+{
+  // TODO(AlexZ): Take all lanugages into an account.
+  featureName.AddString(StringUtf8Multilang::DEFAULT_CODE, newName);
+}
+
+static FeatureID GenerateNewFeatureId(FeatureID const & oldFeatureId)
+{
+  // TODO(AlexZ): Stable & unique features ID generation.
+  static uint32_t newOffset = 0x0effffff;
+  return FeatureID(oldFeatureId.m_mwmId, newOffset++);
+}
+
+void Editor::EditFeatureName(FeatureID const & fid, string const & newName, Index const & index)
+{
+  FeatureType editedFeature;
```

Comment:
FYI: memory operations between ANY two threads may or may not be ordered by a happens-before relation what leads to data races. Memory operations in Android UI threads for the same app are implicitly ordered by a happens-before relation because the second thread starts to work after the first one flushes all operations.

If you're guarantee that Editor will always be called from the UI thread, then ThreadChecker and its macroses are the best choice to check that Editor is used in a thread-safe manner. That's because DECLARE_THREAD_CHECKER() and ASSERT_THREAD_CHECKER() are empty on  Android. On Desktop and iOS UI thread is the same for the whole app lifetime, and DECLARE_THREAD_CHERE() and ASSERT_THREAD_CHECKER() will work fine.